### ISSUE TITLE
Packages endpoint fails for repositories with "github" type

### ISSUE THREAD

**Issue by mmurto** [eclipse-apoapsis/ort-server#1137]
Running an analysis for [mime-types](https://github.com/jshttp/mime-types) results in the server otherwise working, but the packages endpoint failing with the following error:

```
java.lang.IllegalArgumentException: No enum constant org.eclipse.apoapsis.ortserver.api.v1.model.ExtendedRepositoryType.github
	at java.base/java.lang.Enum.valueOf(Enum.java:273)
2024-10-02 12:08:17.162 [eventLoopGroupProxy-4-7] [traceId=fbcf41a7-3598-43a0-973e-9b1e60804275, component=core] level=ERROR StatusPages - Internal Server Error
	at org.eclipse.apoapsis.ortserver.api.v1.mapping.MappingsKt.mapToApi(Mappings.kt:518)
	at org.eclipse.apoapsis.ortserver.api.v1.mapping.MappingsKt.mapToExtendedApi(Mappings.kt:351)
	at org.eclipse.apoapsis.ortserver.api.v1.model.ExtendedRepositoryType.valueOf(VcsInfo.kt)
	at org.eclipse.apoapsis.ortserver.api.v1.mapping.MappingsKt.mapToApi(Mappings.kt:540)
	at org.eclipse.apoapsis.ortserver.core.api.RunsRouteKt$runs$1$6$1$1$pagedResponse$1.invoke(RunsRoute.kt:185)
	at org.eclipse.apoapsis.ortserver.api.v1.mapping.MappingsKt.mapToApi(Mappings.kt:869)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at org.eclipse.apoapsis.ortserver.core.api.RunsRouteKt$runs$1$6$1$1.invokeSuspend(RunsRoute.kt:185)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:101)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:405)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.ktor.server.netty.EventLoopGroupProxy$Companion.create$lambda$1$lambda$0(NettyApplicationEngine.kt:296)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:840)
2024-10-02 12:08:17.163 [eventLoopGroupProxy-4-7] [traceId=fbcf41a7-3598-43a0-973e-9b1e60804275, component=core] level=INFO  Application - 500 Internal Server Error: GET - /api/v1/runs/1/packages in 1728ms
```

The bug can be easily reproduced with #1125 as it has the UI fetch the packages endpoint. Create a fresh Docker compose environment, analyze the project with revision master, only analyzer and reporter enabled, otherwise default settings. After the run, the run overview page should throw the error after failing the packages query.

I haven't thoroughly tested other repositories, but the two that I tried worked as they should without the error.

**Comment by sschuberth** [comment]
Interesting. I wonder where that bogus "github" as a `ExtendedRepositoryType` comes from...

**Comment by mmurto** [comment]
> Interesting. I wonder where that bogus "github" as a `ExtendedRepositoryType` comes from...

Found this from the ort result:

```
    - id: "NPM::js-sdsl:4.4.2"
      purl: "pkg:npm/js-sdsl@4.4.2"
      authors:
      - "ZLY201"
      declared_licenses:
      - "MIT"
      declared_licenses_processed:
        spdx_expression: "MIT"
      description: "javascript standard data structure library which benchmark against\
        \ C++ STL"
      homepage_url: "https://js-sdsl.org"
      binary_artifact:
        url: ""
        hash:
          value: ""
          algorithm: ""
      source_artifact:
        url: "https://registry.npmjs.org/js-sdsl/-/js-sdsl-4.4.2.tgz"
        hash:
          value: "2e3c031b1f47d3aca8b775532e3ebb0818e7f847"
          algorithm: "SHA-1"
      vcs:
        type: "github"
        url: "https://github.com/js-sdsl/js-sdsl.git"
        revision: ""
        path: ""
      vcs_processed:
        type: "github"
        url: "https://github.com/js-sdsl/js-sdsl.git"
        revision: ""
        path: ""
```

**Comment by mmurto** [comment]
And that packages has [this](https://github.com/js-sdsl/js-sdsl/blob/07d24014a3cb69adffa717ee1170824ac5b841c3/package.json#L123-L126). Does ORT read this somewhere?

**Comment by mmurto** [comment]
> And that packages has [this](https://github.com/js-sdsl/js-sdsl/blob/07d24014a3cb69adffa717ee1170824ac5b841c3/package.json#L123-L126). Does ORT read this somewhere?

Apparently it does: https://github.com/oss-review-toolkit/ort/blob/062221e10cd9c4474839b06118a073378995056d/plugins/package-managers/node/src/main/kotlin/utils/NpmSupport.kt#L118-L138

**Comment by mmurto** [comment]
So we probably either need to treat repository type as an arbitrary string, or have [this](https://github.com/oss-review-toolkit/ort/blob/062221e10cd9c4474839b06118a073378995056d/model/src/main/kotlin/VcsType.kt#L72-L75) map to unknown if the type is not known. Not sure, not super familiar with that part of ORT.

**Comment by sschuberth** [comment]
> `type: "github"`

Again interesting. This is something that would / should never occur with plain ORT. So I believe the server should not write this either.

**Comment by mmurto** [comment]
> > `type: "github"`
> 
> Again interesting. This is something that would / should never occur with plain ORT. So I believe the server should not write this either.

I tried analyzing the repo with locally installed ORT (29.1.0-008.sha.4813be3), and the analyzer result from it has the same:

```
      vcs:
        type: "github"
        url: "https://github.com/js-sdsl/js-sdsl.git"
        revision: ""
        path: ""
      vcs_processed:
        type: "github"
        url: "https://github.com/js-sdsl/js-sdsl.git"
        revision: ""
        path: ""
```

**Comment by sschuberth** [comment]
> the analyzer result from it has the same:

OMG, more NPM packages violating the spec... https://github.com/oss-review-toolkit/ort/pull/9235 should work around that.

**Comment by sschuberth** [comment]
> OMG, more NPM packages violating the spec... [oss-review-toolkit/ort#9235](https://github.com/oss-review-toolkit/ort/pull/9235) should work around that.

@mnonnenmacher do you know out of your head if we need to have some DB migration here, to also make existing data work?

**Comment by sschuberth** [comment]
In general, the underlying problem seems to be that

https://github.com/eclipse-apoapsis/ort-server/blob/1d389c227affdb077fffba30927d7d30a583992f/api/v1/model/src/commonMain/kotlin/VcsInfo.kt#L32-L38

is an `enum class`, whereas ORT's `VcsType` [deliberately is not](https://github.com/oss-review-toolkit/ort/blob/58e3c6f0ab0f176b8730a425b8553a3706ebf6ae/model/src/main/kotlin/VcsType.kt#L27-L28):

> This class is not implemented as an enum as constructing from an unknown type should be supported while maintaining that type as the primary alias for the string representation.

**Comment by mnonnenmacher** [comment]
> @mnonnenmacher do you know out of your head if we need to have some DB migration here, to also make existing data work?

If the database now can contain entries with the type column set to "github", then yes, we should make a migration. But is this even possible with the `ExtendedRepositoryType` being an enum?

**Comment by sschuberth** [comment]
> But is this even possible with the `ExtendedRepositoryType` being an enum?

Not as-is, we'd need to makes changes to it. And if that's anyway required, maybe instead of a DB migration the easier fix is to change `ExtendedRepositoryType` *not* to be an enum.

This whole model classes duplication from ORT is very unfortunate, IMO. If not done properly as happened here, it results in errors that would have not occurred if simply using the same model classes.

**Comment by mmurto** [comment]
AFAICT the repository type is [stored in the database correctly](https://github.com/eclipse-apoapsis/ort-server/blob/b1563e4929ea078c6677fe185201e57c44ebf866/dao/src/main/kotlin/tables/ScanResultsTable.kt#L100-L108) as a [RepositoryType](https://github.com/eclipse-apoapsis/ort-server/blob/b1563e4929ea078c6677fe185201e57c44ebf866/model/src/commonMain/kotlin/RepositoryType.kt#L28-L52) but the mapping from database to the API is done to this `ExtendedRepositoryType` (what is it actually?), which causes the crash. This could maybe be fixed by just changing the mapping layer to use something alike the existing `RepositoryType`.

**Comment by sschuberth** [comment]
> the mapping from database to the API is done to this `ExtendedRepositoryType` (what is it actually?)

No Idea. @mnonnenmacher can you help us understand why `ExtendedRepositoryType` is needed in addition to `RepositoryType`?

**Comment by mnonnenmacher** [comment]
> No idea. @mnonnenmacher can you help us understand why ExtendedRepositoryType is needed in addition to RepositoryType?

Better ask @lamppu because she introduced it.

**Comment by sschuberth** [comment]
@mmurto this should have been fixed 8and auto-closed) by https://github.com/eclipse-apoapsis/ort-server/pull/1183, please verify.

**Comment by mmurto** [comment]
> @mmurto this should have been fixed (and auto-closed) by #1183, please verify.

Confirmed.
